File: LeetCode/cpp/102 - Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        vector<vector<int>> res;
        dfs(root, res, 0);
        return res;
    }

    void dfs(TreeNode *root, vector<vector<int>> &res, int cur)
    {
        if (!root)
            return;

        if (res.size() < cur + 1)
            res.push_back({root->val});
        else
            res[cur].push_back(root->val);

        if (root->left)
            dfs(root->left, res, cur + 1);
        if (root->right)
            dfs(root->right, res, cur + 1);
    }
};

File: LeetCode/grind75/102 - Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        res = []
        q = deque()
        q.append(root)

        while q:
            lv = []
            for i in range(len(q)):
                node = q.popleft()
                lv.append(node.val)

                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            res.append(lv)

        return res
